### PR TITLE
build/tools/amebasmart/gnu_utility: fix unexpected report for wifi csi

### PR DESCRIPTION
**Pull request by zhongnuo-tang**
1. Support ch_opt configuration, allow user to choose between legacy/non-legacy for wifi csi
2. Fix num_sub_carrier in csi header mismatch with actual csi len
3. Fix unexpected csi report with ch_opt=1

### CODE REVIEW

**Comment by hiarungupta** [comment]
CSI logs are not coming, using this PR. Please check.

**Comment by zhongnuo-tang** [comment]
> CSI logs are not coming, using this PR. Please check.

This example is for 11n, did you ensure AP's ACK is 11n rate?

**Comment by hiarungupta** [comment]
> > CSI logs are not coming, using this PR. Please check.
> 
> This example is for 11n, did you ensure AP's ACK is 11n rate?

Yes. Checked with ping (ICMP) request/response both packet.  No CSI logs.

**Comment by zhongnuo-tang** [comment]
> > > CSI logs are not coming, using this PR. Please check.
> > 
> > 
> > This example is for 11n, did you ensure AP's ACK is 11n rate?
> 
> Yes. Checked with ping (ICMP) request/response both packet. No CSI logs.
 CSI raw data done!

[CH INFO] csi_sequence = 48,timestamp = 6309328 us, csi data:
[00]0xc6f11d578828abcd
[01]0xd02a08004ce000c8
[02]0x0101800001006045
[03]0x5900510010003800
[04]0x0070000000300000
[05]0x0f000efc0d010000
[06]0x0310080d0b090e04
[07]0xf110f512f913fe12
[08]0xecffec04ec08ee0d
[09]0xfaeff6f2f2f5effa
[10]0x09ef06ee02edfeee
[11]0x10fa10f60ef30cf1
[12]0x09090c070e0110fd
[13]0xfd0e000e030d060b
[14]0xf409f60bf80dfa0e
[15]0xf2fff102f205f207
[16]0xfaf4f7f6f5f9f3fc
[17]0x07f304f201f2fdf2
[18]0x0bfd0bfa0bf709f4

 CSI raw data done!

[CH INFO] csi_sequence = 49,timestamp = 6378904 us, csi data:
[00]0xc6f11d578828abcd
[01]0x982a08004ce000c8
[02]0x0101800001006155
[03]0x5400190010003800
[04]0x0070000000310000
[05]0x05f502f600010000
[06]0x0dfa0cf70af607f5
[07]0x0d060f0210fe0ffb
[08]0x0210060f080d0b09
[09]0xf710f910fd11ff11
[10]0xf108f109f30df40f
[11]0xf1feef01f004f006
[12]0xfaf3f8f5f4f9f1fc
[13]0x04f302f200f0fef1
[14]0x09f809f806f605f3
[15]0x0a000bff09fd09fb
[16]0x0507070608050902
[17]0xfe0a010903080508
[18]0xf706f807fa08fc0

I am able to run this example with the logs as shown.. Please check if your AP's ack is 11n rate..
Thank you

**Comment by hiarungupta** [comment]
> > > > CSI logs are not coming, using this PR. Please check.
> > > 
> > > 
> > > This example is for 11n, did you ensure AP's ACK is 11n rate?
> > 
> > 
> > Yes. Checked with ping (ICMP) request/response both packet. No CSI logs.
> > CSI raw data done!
> 
> [CH INFO] csi_sequence = 48,timestamp = 6309328 us, csi data: [00]0xc6f11d578828abcd [01]0xd02a08004ce000c8 [02]0x0101800001006045 [03]0x5900510010003800 [04]0x0070000000300000 [05]0x0f000efc0d010000 [06]0x0310080d0b090e04 [07]0xf110f512f913fe12 [08]0xecffec04ec08ee0d [09]0xfaeff6f2f2f5effa [10]0x09ef06ee02edfeee [11]0x10fa10f60ef30cf1 [12]0x09090c070e0110fd [13]0xfd0e000e030d060b [14]0xf409f60bf80dfa0e [15]0xf2fff102f205f207 [16]0xfaf4f7f6f5f9f3fc [17]0x07f304f201f2fdf2 [18]0x0bfd0bfa0bf709f4
> 
> CSI raw data done!
> 
> [CH INFO] csi_sequence = 49,timestamp = 6378904 us, csi data: [00]0xc6f11d578828abcd [01]0x982a08004ce000c8 [02]0x0101800001006155 [03]0x5400190010003800 [04]0x0070000000310000 [05]0x05f502f600010000 [06]0x0dfa0cf70af607f5 [07]0x0d060f0210fe0ffb [08]0x0210060f080d0b09 [09]0xf710f910fd11ff11 [10]0xf108f109f30df40f [11]0xf1feef01f004f006 [12]0xfaf3f8f5f4f9f1fc [13]0x04f302f200f0fef1 [14]0x09f809f806f605f3 [15]0x0a000bff09fd09fb [16]0x0507070608050902 [17]0xfe0a010903080508 [18]0xf706f807fa08fc0
> 
> I am able to run this example with the logs as shown.. Please check if your AP's ack is 11n rate.. Thank you

CSI Logs looks fine for 802.11n.

**Comment by zhongnuo-tang** [comment]
> > > > > CSI logs are not coming, using this PR. Please check.
> > > > 
> > > > 
> > > > This example is for 11n, did you ensure AP's ACK is 11n rate?
> > > 
> > > 
> > > Yes. Checked with ping (ICMP) request/response both packet. No CSI logs.
> > > CSI raw data done!
> > 
> > 
> > [CH INFO] csi_sequence = 48,timestamp = 6309328 us, csi data: [00]0xc6f11d578828abcd [01]0xd02a08004ce000c8 [02]0x0101800001006045 [03]0x5900510010003800 [04]0x0070000000300000 [05]0x0f000efc0d010000 [06]0x0310080d0b090e04 [07]0xf110f512f913fe12 [08]0xecffec04ec08ee0d [09]0xfaeff6f2f2f5effa [10]0x09ef06ee02edfeee [11]0x10fa10f60ef30cf1 [12]0x09090c070e0110fd [13]0xfd0e000e030d060b [14]0xf409f60bf80dfa0e [15]0xf2fff102f205f207 [16]0xfaf4f7f6f5f9f3fc [17]0x07f304f201f2fdf2 [18]0x0bfd0bfa0bf709f4
> > CSI raw data done!
> > [CH INFO] csi_sequence = 49,timestamp = 6378904 us, csi data: [00]0xc6f11d578828abcd [01]0x982a08004ce000c8 [02]0x0101800001006155 [03]0x5400190010003800 [04]0x0070000000310000 [05]0x05f502f600010000 [06]0x0dfa0cf70af607f5 [07]0x0d060f0210fe0ffb [08]0x0210060f080d0b09 [09]0xf710f910fd11ff11 [10]0xf108f109f30df40f [11]0xf1feef01f004f006 [12]0xfaf3f8f5f4f9f1fc [13]0x04f302f200f0fef1 [14]0x09f809f806f605f3 [15]0x0a000bff09fd09fb [16]0x0507070608050902 [17]0xfe0a010903080508 [18]0xf706f807fa08fc0
> > I am able to run this example with the logs as shown.. Please check if your AP's ack is 11n rate.. Thank you
> 
> CSI Logs looks fine for 802.11n.

Do you need other confirmations?

**Comment by hiarungupta** [comment]
> > > > > > CSI logs are not coming, using this PR. Please check.
> > > > > 
> > > > > 
> > > > > This example is for 11n, did you ensure AP's ACK is 11n rate?
> > > > 
> > > > 
> > > > Yes. Checked with ping (ICMP) request/response both packet. No CSI logs.
> > > > CSI raw data done!
> > > 
> > > 
> > > [CH INFO] csi_sequence = 48,timestamp = 6309328 us, csi data: [00]0xc6f11d578828abcd [01]0xd02a08004ce000c8 [02]0x0101800001006045 [03]0x5900510010003800 [04]0x0070000000300000 [05]0x0f000efc0d010000 [06]0x0310080d0b090e04 [07]0xf110f512f913fe12 [08]0xecffec04ec08ee0d [09]0xfaeff6f2f2f5effa [10]0x09ef06ee02edfeee [11]0x10fa10f60ef30cf1 [12]0x09090c070e0110fd [13]0xfd0e000e030d060b [14]0xf409f60bf80dfa0e [15]0xf2fff102f205f207 [16]0xfaf4f7f6f5f9f3fc [17]0x07f304f201f2fdf2 [18]0x0bfd0bfa0bf709f4
> > > CSI raw data done!
> > > [CH INFO] csi_sequence = 49,timestamp = 6378904 us, csi data: [00]0xc6f11d578828abcd [01]0x982a08004ce000c8 [02]0x0101800001006155 [03]0x5400190010003800 [04]0x0070000000310000 [05]0x05f502f600010000 [06]0x0dfa0cf70af607f5 [07]0x0d060f0210fe0ffb [08]0x0210060f080d0b09 [09]0xf710f910fd11ff11 [10]0xf108f109f30df40f [11]0xf1feef01f004f006 [12]0xfaf3f8f5f4f9f1fc [13]0x04f302f200f0fef1 [14]0x09f809f806f605f3 [15]0x0a000bff09fd09fb [16]0x0507070608050902 [17]0xfe0a010903080508 [18]0xf706f807fa08fc0
> > > I am able to run this example with the logs as shown.. Please check if your AP's ack is 11n rate.. Thank you
> > 
> > 
> > CSI Logs looks fine for 802.11n.
> 
> Do you need other confirmations?

Please check and confirm, old binary changes included in new binary. As Logs are being come with Old binary for 11n but not with new binary.

**Comment by hiarungupta** [comment]
Updated PR is looks fine. CSI Logs are being come.

**Review by sunghan-chang — COMMENTED** [review]
I gonna merge this. But before merge, could you split the commit as two?
1. csi feature changes (in build and os folders)
2. csi app changes (in app folder)

**Comment by zhongnuo-tang** [comment]
> I gonna merge this. But before merge, could you split the commit as two?
> 
> 1. csi feature changes (in build and os folders)
> 2. csi app changes (in app folder)

I have split into two commits, please check. Thank you

File: os/include/tinyara/wifi/rtk/wifi_structures.h [hunk 1]
```diff
@@ -134,17 +134,39 @@ typedef struct rtw_scan_handler_result {
   * @brief  The structure is used to describe the cfg parameters used for csi report,
   * @note  The mac_addr if not specified, the default value must be 0.
   */
-#ifdef CONFIG_ARCH_CHIP_RTL8720E
+#if CONFIG_ARCH_CHIP_RTL8720E || CONFIG_ARCH_CHIP_RTL8730E
```
Comment (sunghan-chang) [inline os/include/tinyara/wifi/rtk/wifi_structures.h:137]: 2nd commit for app includes changes on the os. Please move this into the 1st commit.

Comment (zhongnuo-tang) [reply os/include/tinyara/wifi/rtk/wifi_structures.h:137]: Changed accordingly

**Comment by zhongnuo-tang** [comment]
Dear @sunghan-chang,
There is a merge conflict, I have raised another PR for this fix https://github.com/Samsung/TizenRT/pull/6140. Please help to check there and I will abandon this PR. 
Thank you

**Pull request closed without merging**